Overhauls blocking + a couple of new toys to block with.

Rukofamicom opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

### Core blocking changes
* **Block levels are gone** - All items that can block will block the tiles directly and diagonally in front of you. Attacks from the sides or rear diagonals will bypass blocking attempts. Walk/Run no longer has any impact on your ability to block.
* **Blocking with a weapon has a 2 second cooldown**  unless you were blocking projectiles, because the weapons that do this need to be able to close gaps. Blocking with a weapon also requires you to have less than 45 stamina damage. 
* Blocking with a shield still has no cooldown and they work from the off-hand. If you have more than 45 stamina damage after blocking with a shield it will be sent flying from your hand unless it is an energy shield. If you glued the shield to your hand to prevent dropping, it will be shattered early. 
* Blocking stamina is no longer limb based
* Added three new blocking flags: 
  * BLOCKING_COUNTERATTACK which is an upgraded version of BLOCKING_NASTY, and allows you to counter-attack even if the opponent isn't barehanded but only at a 50% chance
  * BLOCKING_UNBALANCE which gives a 20% chance to trip your opponent and prevent them from attacking for 2x the normal melee attack cooldown
  * BLOCKING_UNBLOCKABLE which will enable weapons to completely bypass blocking *unless* they are blocked by a weapon that also shares this trait. This trait is reserved for energy swords (including chaplain variant) and energy shields. Only an energy saber or shield can block an energy saber. 
* TRAIT_NOBLOCK now applies to shields as well. If you can't block because you took drugs you simply can't block. Shields are no longer a way around this. 
* Blocking an incoming melee attack has entirely new calculations that follow the same spirit as the old ones but with less extreme outcomes. Sharp and non-brute weapons are easier to block than blunt ones. 
tweak: blocking stamina is no longer limb based
* Simplemobs no longer do 2x damage when blocked, unless they have an already existing special variable set which marks them as especially dangerous for players trying to block. If they have this variable set they do 5x blocking damage and are probably hitting the block cap unless you're using something with exceptional blocking ability

### Items that have had their blocking ability dramatically changed
* The magical barrier nullrod staff incurs no stamina damage when blocking and has been given BLOCK_UNBLOCKABLE and BLOCKING_UNBALANCE since the entire point of this choice is magical defense and it only does 5 damage. You are required to wield the staff in both hands to activate this barrier now, however. This also means you can use this as your holy weapon and not have an eternal visual barrier even when it isn't active. 
* The rod of Asclepius incurs no stamina damage when blocking and has the flags for BLOCK_UNBALANCE and BLOCK_PROJECTILE. The rod only gains the ability block after activation which gives the wielder pacifism so it seemed appropriate to make this an extremely powerful blocking tool for someone who can never drop it and never fight back. This is the only item that can block from the off-hand while not being a shield. 
* Chaplain energy sword has gained BLOCKING_PROJECTILE and BLOCKING_UNBLOCKABLE
* Regular energy swords have gained BLOCKING_PROJECTILE and BLOCKING_UNBLOCKABLE
* Both botany and chaplain scythes have been upgraded from BLOCKING_NASTY to BLOCKING_COUNTERATTACK. Both types of scythes now do 18 damage and have SHARP_DISMEMBER sharpness as well, making them fairly formidable weapons. 
* The chaplain bo staff has been given parity with the newly implemented bo staff of the same appearance and it can now be two-handed. Both deal 10 damage if one-handed and 14 if two-handed, as well as sharing the BLOCKING_COUNTERATTACK and BLOCKING_UNBALANCE flags making them excellent weapons with a focus on defense. 
* Knives and all subtypes of knives can now block with the BLOCKING_NASTY flag, but aside from this they are terrible at blocking due to a block_power of 0
* The captain's sabre has been given the BLOCKING_COUNTERATTACK flag, alongside the newly added carbon fiber sabre
* Sabre sheathes now have moderate blocking ability, and a very specific interaction with sabers that causes them to sheathe and steal a saber if they block a hit from one while being empty.
* Mops have the ability to block and the flag for BLOCKING_UNBALANCE, but have a block_power of 0
* Spears have the ability to block when wielded now and have the flags for BLOCKING_COUNTERATTACK and BLOCKING_UNBALANCE making them one of the most well-rounded weapons. 
* Teleprods can now block and have the flag for BLOCKING_COUNTERATTACK
* The ordinary wizard staff (not the magical ones) as well as subtypes (broom and stick) have 50 blocking power and the BLOCKING_UNBALANCE flag. The staff has also been given 12 damage and the other two 5 damage. They are still awful and you have very little reason to use them outside of memes, even as a non-wizard. 
* Proto-kinetic crushers can now block, though they ahve a block power of 0
* Ghost sword's block power is determined by the number of ghosts, reaching 100 block power with five ghosts and making it cost no stamina to block with. It has no other notable changes. 

### The new toys
* I've implemented the carbon fiber sabre lost to time from #6125, although my version have very different stats and are available in all security lockers instead of through cargo. These swords have the same stats as the captain's sabre except for having less armor penetration and ordinary sharpness. 
* I've re-implemented the bo staff, an item which existed in code but was unobtainable without admin intervention. I've tuned this item to be appropriate to craftable station weaponry with stats very similar to the common spear. Compared to the spear it is blunt, but the wielder will take less stamina damage when blocking with it.

### The stuff I'm legally required to point out
* I have updated the combat info for items with blocking capabilities
* I have more or less completely removed the bo staff as it previously existed in code, and even relocated it to a more appropriate file. It was previously part of the sleeping carp file and clearly balanced to be an antagonist weapon along the same power level. It is now a weapon craftable from wood and balanced accordingly. 
* I changed a lot of return 1 and return 0 to TRUE/FALSE accordingly
* removed all lingering remnants of an unused variable called "final_block_chance"
* SORD and UNREAL SORD have had their shitcode cleaned up a bit and been made to match each other. They still suck absolute ass because that's the point. 
* Because the mime sabre has lost a lot of what made it worth getting with blocking nerfs, its damage has been buffed to 25. It is still largely inferior to energy sabers, but is at least a viable flavor option with this.
* Wizard hammers have lost their ability to block because they only had this ability when walking before and I don't believe they need the ability to block given the wizards' power level. 
* Shields no longer drop debris or scrap when broken
* An unused spear called "The Grey Tide" has been removed from the code due to being unused and probably undesirable for adminbus as well. 
* Bone spears have received dramatically increased armor penetration so they can actually get through insane miner armor
* Replica Katanas now do 15 damage and are a bit sharper. They aren't "toy" katanas or they wouldn't have been sharp in the first place, so I made them a middle-ground weapon that's easily accessible. They have the same blocking flags as the other Katanas as well. 
* Energy shields now work again. At some point they were broken twice over - they no longer recharged, but also didn't need to recharge because you could just immediately reactivate them after shatter to block another hit.

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

This has been explained mostly point for point above already so I'm going to keep it simple in this section: Blocking was a very opaque and complicated system, and it is still somewhat complex but no longer feels extremely inconsistent for players who aren't intimately familiar with the mechanics. 

It's much simpler to look at blocking as having two main flavors:
* If you're holding a weapon that can block, it will block once every few seconds, and if you're attacking someone with a weapon that can block they will only be able to block once every few seconds.
* If you're holding a shield, you will automatically block with it for as long as your stamina (and shield durability) lasts even if it isn't in your active hand. If you're attacking someone with a shield, you know it will be knocked away after a few hits and won't protect them forever.
* If an item (shield or weapon) has certain properties, those properties will happen automatically without further input or skill required by the player holding them. 

In addition to the above, blocking was either completely absent from combat or totally dominated it depending on exactly what was being fought with, and when blocking dominated one side of combat it usually felt very unfair for the other half of combat because blocking would enable players to survive four or five times as much incoming damage as they would otherwise be able to. On the balance side of things, this PR intends to make blocking much more prevelant by adding it to many weapons that didn't previously have it, but also prevent blocking mechanics from totally stonewalling someone by dramatically reducing shield durability/stamina reduction and introducing the cooldown. 


<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding.
-->

## Testing Photographs and Procedure
<!-- Include any screenshots/videos/debugging steps of the modified code functioning successfully, ideally including edge cases. -->
<details>
<summary>Screenshots&Videos</summary>

![dreamseeker_RsXOOjdYn9](https://github.com/user-attachments/assets/95838205-4188-444d-80f3-005002758e80)

![dreamseeker_83E9BiXpsF](https://github.com/user-attachments/assets/a18fc101-2da2-405e-9d9a-e57ffa334b38)

![dreamseeker_Jwq7YVSTjB](https://github.com/user-attachments/assets/931e295f-ec44-4103-807a-d357fddf1151)

Counterattack and unbalance in play on the new bo-staff
![dreamseeker_ylS3W1Eq8M](https://github.com/user-attachments/assets/47b683e2-094e-430f-80c6-2896f58e0dbb)

Sheathing and stealing a sabre
<img width="548" height="59" alt="image" src="https://github.com/user-attachments/assets/1149f0fd-6b1a-44fd-ac31-4ca32650c355" />

Blocking (notably not reflecting) projectiles with a saber, and also showing that they can still be shot from the sides
![dreamseeker_kWkobbex79](https://github.com/user-attachments/assets/2eab585c-53cf-48a3-8803-73eeca20426a)

</details>

## Changelog
:cl: Powerfulbacon (sprites), Rukofamicom

del: Removed block levels from all items. Items that can block will block the tiles directly and diagonally in front of you. Attacks from the sides or rear diagonals will bypass blocking attempts. Walk/Run no longer has any impact on your ability to block.
balance: Blocking with a weapon now has a two second cooldown unless you were blocking projectiles, and it also triggers your melee attack cooldown slowing your next attack. Blocking with a weapon will also fail if you have more than 45 stamina damage. 
balance: Blocking with shields now results in faster stamina damage buildup, and the shield will be sent flying from your hand at more than 45 stamina damage
balance: Blocking an incoming melee attack has entirely new calculations that follow the same spirit as the old ones but with less extreme outcomes. Sharp and non-brute weapons are easier to block than blunt ones. 
tweak: blocking stamina is no longer limb based
tweak: Added or changed blocking capabilities and/or damage stats of the following weapons. Examine them in-game for more details: Many nullrod options, Rod of Asclepius (activated), energy swords, scythes, spears, proto-kinetic accelerator, mime traitor baguette, mops, teleprods, ghost sword. 
add: three new blocking flags that will be revealed on examining weapons. The flags are BLOCKING_COUNTERATTACK which does what it sounds like, BLOCKING_UNBALANCE for long weapons which trips your opponent and delays their next attack, and BLOCKING_UNBLOCKABLE which can only be blocked by itself and is reserved for magical or energy weapons and energy shields. 
add: Adds the Carbon Fiber Saber to security lockers scattered across the station. This item is nearly identical to the captain's saber, but has much less armor penetration and normal sharpness.
add: "Adds" a non nullrod bo staff, a two-handed weapon craftable from wood that excels at blocking attacks and tripping up your opponent. This weapon is comparable to a spear in damage, but is blunt and incurs less stamina damage while blocking. 
fix: Fixes energy shield in-hand sprites so that they are now visible in player hands again
fix: Fixes energy shields functionality as well - they would never recharge, but this may have gone unnoticed because they also weren't required to charge before being deployed again. 
tweak: Energy shields are now much cheaper (5 TC) and available in regular uplinks instead of only nuclear operative uplinks. This extends to the bananium shield being cheaper for clowns as well. 
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Rukofamicom committed:
Variable renames

Rukofamicom committed:
Merge branch 'blocking-cooldown' of https://github.com/Rukofamicom/BeeStation-Hornet into blocking-cooldown

Rukofamicom committed:
shields with NO_DROP call shatter() if stam is too high

Rukofamicom committed:
SORDs can't block

Well they sort of can, but it's a really bad idea to try.

Rukofamicom committed:
Beepsky smash no longer gives NO_BLOCK

Rukofamicom committed:
removes blockbreak() from humans

redundant proc better handled with a COOLDOWN macro

Rukofamicom committed:
Supermatter sword comment

Rukofamicom committed:
Defines blocking cooldown under combat defines

Rukofamicom committed:
Updates simplemob variable usage

Rukofamicom committed:
Core blocking changes

Rukofamicom committed:
fix block angles

Rukofamicom committed:
Debugging

I can't prevent a block from within on_block, has to be in hit_reaction

Rukofamicom committed:
Three new flags

Rukofamicom committed:
Energy weapons

block projectiles + unblockable flags

Rukofamicom committed:
fixes SORD, gives it active blocking

Rukofamicom committed:
Sticks and Scythes

Wizard and Chaplain staves, Chaplain Scythe, Bo staves

Rukofamicom committed:
shields

Rukofamicom committed:
replica katana

Is now a weak katana, but a usable weapon

Rukofamicom committed:
Misc.

Rukofamicom committed:
shield tweak and warning

Rukofamicom committed:
removes unused variable "final_block_chance"

Rukofamicom committed:
Carbon Fiber and other sabre shenanigans

Rukofamicom committed:
spears, knives and kinetic crusher

Rukofamicom committed:
fixes as I gather testing evidence

Rukofamicom committed:
undo basemap

Rukofamicom commented on code/game/objects/items.dm:
```diff
@@ -639,98 +643,110 @@ GLOBAL_VAR_INIT(rpg_loot_items, FALSE)
+			attackforce = attackforce * 5 //You can probably only block them once or twice at most because of stamina and/or shield damage
+
+	//We are ready to deal stamina damage to our owner
+	owner.apply_damage(min(attackforce, 60), STAMINA, blocked = block_power)
```
I forgot to come back and improve this until I'd already posted this. I need to make certain this can't stamcrit

Rukofamicom commented on code/game/objects/items.dm:
```diff
@@ -832,7 +848,7 @@ GLOBAL_VAR_INIT(rpg_loot_items, FALSE)
 	return TRUE
 
-//the mob M is attempting to equip this item into the slot passed through as 'slot'. Return 1 if it can do this and 0 if it can't.
+//the mob M is attempting to equip this item into the slot passed through as 'slot'. return TRUE if it can do this and 0 if it can't.
```
```suggestion
//the mob M is attempting to equip this item into the slot passed through as 'slot'. return TRUE if it can do this and FALSE if it can't.
```

Rukofamicom commented on code/game/objects/items/food/misc.dm:
```diff
@@ -180,8 +180,8 @@
 	throwforce = 15
-	block_level = 2
-	block_upgrade_walk = TRUE
+	canblock = TRUE
```
Give this blocking flags

Rukofamicom commented on code/game/objects/items/his_grace.dm:
```diff
@@ -15,7 +15,7 @@
 	item_flags = ISWEAPON
 	force = 12
-	block_upgrade_walk = TRUE
+
```
This should be considered for blocking flags

Rukofamicom committed:
I forgot the mop

Rukofamicom commented on code/game/objects/items/pitchfork.dm:
```diff
@@ -6,8 +6,8 @@
 	throwforce = 15
-	block_level = 1
-	block_upgrade_walk = TRUE
+	canblock = TRUE
```
flags

Rukofamicom replied:
Actually this is yet another completely unused and unavailable item so I'm just removing it.

Rukofamicom committed:
Teleprod

Rukofamicom commented on code/modules/antagonists/clock_cult/enchantments/blocking.dm:
```diff
@@ -3,6 +3,5 @@
 /datum/component/enchantment/blocking/apply_effect(obj/item/target)
 	examine_description = "It has been blessed with the gift of blocking."
-	target.block_level = min(level - 1, 1)
+	target.canblock = TRUE
```
flags probably. Maybe not actually since this would be cooler if they didn't require active hand.

Rukofamicom commented on code/modules/mining/equipment/kinetic_crusher.dm:
```diff
@@ -31,6 +31,10 @@
+	canblock = TRUE
+	block_flags = BLOCKING_ACTIVE
+	block_power = 0
+
```
Make this only able to block when wielded like all the other wielded blocking tools

Rukofamicom commented on code/modules/mob/living/carbon/human/species_types/shadowpeople.dm:
```diff
@@ -251,7 +251,11 @@
 	force = 25
+
+	block_power = 50
+	canblock = TRUE
```
Revert this actually, this doesn't need the ability to block. I missed it when I pulled back from the other armblades.

Rukofamicom committed:
Merge remote-tracking branch 'upstream/master' into blocking-cooldown

Rukofamicom committed:
no stamcrit when blocking

Rukofamicom committed:
shield consistency

Rukofamicom committed:
FALSE not 0

Rukofamicom committed:
Missing flags, missing wield checks, revert nightmare

Rukofamicom committed:
Merge branch 'blocking-cooldown' of https://github.com/Rukofamicom/BeeStation-Hornet into blocking-cooldown

Rukofamicom committed:
Removes pitchfork

Rukofamicom committed:
Revert "Removes pitchfork"

This reverts commit 6f74e8916a5b1a14ad4d80532ad0019e00d4db4f.

Rukofamicom committed:
Fixes pitchfork

Nevermind the linter throws a shitfit if I try to remove it and I'm not fucking with that.

Rukofamicom committed:
Polish, fixes and flavor messages

PowerfulBacon commented on code/game/objects/items.dm:
```diff
@@ -643,98 +647,112 @@ GLOBAL_VAR_INIT(rpg_loot_items, FALSE)
+			attackforce = max(I.w_class * 3, attackforce * 1.5)
+
+		//Is it a weapon especially adept at counterattacks? If so we roll for one
+		if(block_flags & BLOCKING_COUNTERATTACK && prob(50))
```
Add proper brackets to bitflag statements as they are dangerous
```suggestion
		if((block_flags & BLOCKING_COUNTERATTACK) && prob(50))
```

Rukofamicom replied:
Will do

Rukofamicom replied:
Went through and contained all of them

Rukofamicom committed:
Merge branch 'BeeStation:master' into blocking-cooldown

Rukofamicom committed:
whoops

Rukofamicom committed:
Merge branch 'blocking-cooldown' of https://github.com/Rukofamicom/BeeStation-Hornet into blocking-cooldown

Rukofamicom committed:
contain the dangerous flags

Rukofamicom committed:
fix stun baton

Rukofamicom committed:
Tweaks after first test

Rukofamicom committed:
cheaper energy shields

Rukofamicom committed:
Fixes shield icons

Rukofamicom committed:
updates ashwalker camp

Rukofamicom committed:
Adds backsprites for shields that now fit on backs

Rukofamicom committed:
Fix shields by using AND not OR

Rukofamicom committed:
Update items.dm

Rukofamicom committed:
Fix chaplain bo staff

Rukofamicom committed:
block_power tweaks after testing, fixes combat info

Rukofamicom committed:
Touches up chaplain holy staff, you can now turn the barrier on and off.

Rukofamicom committed:
Merge remote-tracking branch 'upstream/master' into blocking-cooldown

Rukofamicom committed:
Fixes energy shields

Non functional recharging capabilities have been fixed and the sound that plays when the shield is charged is now audible.

Rukofamicom committed:
Hanzo steel brought down to match the other station katanas

mrmanlikesbt reviewed (approved):
Ruko talked to me about the concerns I had and I'm happy to let this go through now.

@PowerfulBacon I remember you saying something about this PR, you fine with this getting merged?

Rukofamicom committed:
Improve consistency of maximum blocking stamina damage

The absolute maximum under the stars aligning of stamina damage someone can take is now 69, if they had 34 when they got hit and took the maximum stamina output of 35
